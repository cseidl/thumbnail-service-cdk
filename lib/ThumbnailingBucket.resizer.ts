import AWS, { S3 } from "aws-sdk";
import {
    S3Event
} from "aws-lambda";
import sharp from "sharp";

// extracts the extension from a path and returns [path without extension, extension]
function pathParts(name: string) {
    const groups = /^(.+)\.(.+?)$/.exec(name);
    if (!groups) {
        return [name, ""];
    }
    const [, base, ext] = groups;
    return [base, ext];
}

export const handler = async (
    event: S3Event
): Promise<void> => {
    AWS.config.logger = console;

    const destBucket = process.env.DEST_BUCKET!;
    console.log(JSON.stringify(event));
    const s3 = new S3();

    for (const s3Event of event.Records) {
        const key =  decodeURIComponent(s3Event.s3.object.key.replace(/\+/g, ' ')); // space is encoded as +
        const obj = await s3.getObject({
            Bucket: s3Event.s3.bucket.name,
            Key: key,
        }).promise();

        const [base, ext] = pathParts(key);

        const list = [];

        for (let size of [50, 100, 200]) {
            const out = await sharp(obj.Body! as Buffer).resize({
                height: size,
                width: size
            }).toBuffer({ resolveWithObject: true });
            const destKey = `${base}-${size}x${size}.${out.info.format}`;
            console.log(`${key}: Creating ${destKey}`);
            await s3.putObject({
                Bucket: destBucket,
                Key: destKey,
                Body: out.data
            }).promise();
            list.push({
                key: destKey,
                url: `https://${destBucket}.s3.amazonaws.com/${destKey}`,
                width: size,
                height: size
            });
        }

        console.log(`${key}: Creating thumbnails.json`);
        await s3.putObject({
            Bucket: destBucket,
            Key: `${key}.thumbnails.json`,
            Body: JSON.stringify(list),
            ContentType: "application/json"
        }).promise();
    }
};
