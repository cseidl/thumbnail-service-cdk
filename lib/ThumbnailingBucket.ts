import { Construct, Duration, RemovalPolicy } from "@aws-cdk/core";
import * as s3 from '@aws-cdk/aws-s3';
import * as s3notify from '@aws-cdk/aws-s3-notifications';
import * as lambda_node from "@aws-cdk/aws-lambda-nodejs";
import * as lambda from "@aws-cdk/aws-lambda";

export interface ThumbnailingBucketProps {
    name?: string;
}

export class ThumbnailingBucket extends Construct {
    sourceBucket: s3.Bucket;
    destBucket: s3.Bucket;
    func: lambda_node.NodejsFunction;

    constructor(scope: Construct, id: string, props: ThumbnailingBucketProps = {}) {
        super(scope, id);

        this.sourceBucket = new s3.Bucket(this, 'sourceBucket', {
            bucketName: props.name,
            removalPolicy: RemovalPolicy.DESTROY
        });
        this.destBucket = new s3.Bucket(this, 'destBucket', {
            bucketName: `${this.sourceBucket.bucketName}-thumbs`,
            removalPolicy: RemovalPolicy.DESTROY
        });
        this.func = new lambda_node.NodejsFunction(this, 'resizer', {
            bundling: {
                forceDockerBundling: true,
                nodeModules: ['sharp'],
            },
            timeout: Duration.minutes(1),
            memorySize: 256,
            functionName: `${props.name}-resizer`,
            runtime: lambda.Runtime.NODEJS_14_X,
            environment: {
                DEST_BUCKET: this.destBucket.bucketName
            },
        });

        [".jpg", ".jpeg", ".png"].forEach(suffix =>
            this.sourceBucket.addObjectCreatedNotification(
                new s3notify.LambdaDestination(this.func), { suffix: suffix }));
        this.sourceBucket.grantRead(this.func);
        this.destBucket.grantWrite(this.func);
    }
}