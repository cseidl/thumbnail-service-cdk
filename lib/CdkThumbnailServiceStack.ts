import * as cdk from '@aws-cdk/core';
import { CfnOutput } from "@aws-cdk/core";
import { ThumbnailingBucket } from "./ThumbnailingBucket";

export class CdkThumbnailServiceStack extends cdk.Stack {
    constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        const thumbnailer = new ThumbnailingBucket(this, "Thumbnailer", {
            name: "lukstei-pictures2"
        });

        new CfnOutput(this, "SourceBucketUrl", {
            description: "Source Bucket",
            value: `https://s3.console.aws.amazon.com/s3/buckets/${thumbnailer.sourceBucket.bucketName}`
        })
        new CfnOutput(this, "DestBucketUrl", {
            description: "Destination Bucket",
            value: `https://s3.console.aws.amazon.com/s3/buckets/${thumbnailer.destBucket.bucketName}`
        })
        new CfnOutput(this, "LambdaOut", {
            description: "Lambda ARN",
            value: thumbnailer.func.functionArn
        })
    }
}
