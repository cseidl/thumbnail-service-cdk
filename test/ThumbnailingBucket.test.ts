import { Stack } from '@aws-cdk/core';
import { ThumbnailingBucket } from "../lib/ThumbnailingBucket";
import '@aws-cdk/assert/jest';
import { SynthUtils } from "@aws-cdk/assert";

describe('ThumbnailingBucket', () => {
    const stack = new Stack();
    new ThumbnailingBucket(stack, 'Subject', {
        name: "blog-pictures"
    });
    it("defines exactly two buckets", () => {
        expect(stack).toCountResources('AWS::S3::Bucket', 2);
    });
    it("defines exactly two buckets", () => {
        expect(stack).toCountResources('AWS::Lambda::Function', 1+1); // 1 is implicitly created by the notification handler
    });
    it("sets the correct policies", () => {
        expect(SynthUtils.subset(stack, {resourceTypes: ["AWS::IAM::Policy"]})).toMatchSnapshot();
    });
});