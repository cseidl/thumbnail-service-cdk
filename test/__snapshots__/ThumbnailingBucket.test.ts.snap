// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThumbnailingBucket sets the correct policies 1`] = `
Object {
  "Parameters": Object {
    "AssetParametersa2f31fca9c5c89d231767354edd7a1cdf59f6ac115097449d573480adcd87893ArtifactHashD5EA1BD9": Object {
      "Description": "Artifact hash for asset \\"a2f31fca9c5c89d231767354edd7a1cdf59f6ac115097449d573480adcd87893\\"",
      "Type": "String",
    },
    "AssetParametersa2f31fca9c5c89d231767354edd7a1cdf59f6ac115097449d573480adcd87893S3Bucket58130362": Object {
      "Description": "S3 bucket for asset \\"a2f31fca9c5c89d231767354edd7a1cdf59f6ac115097449d573480adcd87893\\"",
      "Type": "String",
    },
    "AssetParametersa2f31fca9c5c89d231767354edd7a1cdf59f6ac115097449d573480adcd87893S3VersionKey427CAB8F": Object {
      "Description": "S3 key for asset version \\"a2f31fca9c5c89d231767354edd7a1cdf59f6ac115097449d573480adcd87893\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:PutBucketNotification",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleDefaultPolicy2CF63D36",
        "Roles": Array [
          Object {
            "Ref": "BucketNotificationsHandler050a0587b7544547bf325f094a3db834RoleB6FB88EC",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "SubjectresizerServiceRoleDefaultPolicy12948A57": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "SubjectsourceBucketDE59AEC9",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "SubjectsourceBucketDE59AEC9",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "SubjectdestBucketF0E8C51E",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "SubjectdestBucketF0E8C51E",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "SubjectresizerServiceRoleDefaultPolicy12948A57",
        "Roles": Array [
          Object {
            "Ref": "SubjectresizerServiceRole37921AD8",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
